File: doc/legacy_ts_syntax/sdk/step.ts
import { emitEvent, log } from "./events";

export async function step(label: string, fn: () => Promise<void>): Promise<void> {
  const ts = new Date().toISOString();
  log("step", label);
  emitEvent({ type: "step", step: label, status: "started", timestamp: ts });

  try {
    await fn();
    const doneTs = new Date().toISOString();
    log("step", `${label} done`);
    emitEvent({ type: "step", step: label, status: "completed", timestamp: doneTs });
  } catch (e: any) {
    const failTs = new Date().toISOString();
    log("step", `${label} failed: ${String(e)}`);
    emitEvent({ type: "step", step: label, status: "failed", timestamp: failTs, error: String(e) });
    throw e;
  }
}

File: sdk/step.go
package sdk

import (
	"fmt"
	"time"
)

func Step(label string, fn func()) {
	ts := time.Now().Format(time.RFC3339)
	logMsg("step", label)
	emitEvent(Event{Type: "step", Step: label, Status: "started", Timestamp: ts})

	defer func() {
		if r := recover(); r != nil {
			failTs := time.Now().Format(time.RFC3339)
			logMsg("step", fmt.Sprintf("%s failed: %v", label, r))
			emitEvent(Event{Type: "step", Step: label, Status: "failed", Timestamp: failTs, Error: fmt.Sprint(r)})
			panic(r)
		}
	}()

	fn()

	doneTs := time.Now().Format(time.RFC3339)
	logMsg("step", label+" done")
	emitEvent(Event{Type: "step", Step: label, Status: "completed", Timestamp: doneTs})
}
